lib/ReadsUtils/ReadsUtilsImpl.py

```diff
@@ -642,6 +668,223 @@ def get_fq_stats(self, reads_object, file_path):
             reads_object[key] = ea_stats_dict[key]
         return reads_object
 
+
+    def _get_staging_file_path(self, token_user, upload_file_name):
+        """
+        _get_staging_file_path: return staging area file path
+
+        directory pattern: /data/bulk/user_name/file_name
+
+        """
+        return '/data/bulk/%s/%s' % (token_user, upload_file_name)
+
+
+    def _download_staging_file(self, token_user, staging_file_name):
+        """
+        _download_staging_file: download staging file to scratch
+
+        return: file path of downloaded staging file
+
+        """
+
+        staging_file_path = self._get_staging_file_path(token_user, staging_file_name)
+
+        dstdir = os.path.join(self.scratch, 'tmp')
+        if not os.path.exists(dstdir):
+            os.makedirs(dstdir)
+        shutil.copy2(staging_file_path, dstdir)
+        copy_file_path = os.path.join(dstdir, staging_file_path)
+
+        return copy_file_path
+
+
+    def _download_web_file(self, file_url, download_type, rev_file=False):
+        """
+        _download_web_file: download reads source file from web
+
+        file_url: file URL
+        download_type: one of ['Direct Download', 'FTP', 'DropBox', 'Google Drive']
+        rev_file: optional, default as False. Set to True if file_url is for rev_file
+
+        return: file path of downloaded web file
+
+        """
+
+        # prepare local copy file path for copy
+        tmp_file_name = 'tmp_rev_fastq.fastq' if rev_file else 'tmp_fwd_fastq.fastq'
+        dstdir = os.path.join(self.scratch, 'tmp')
+        if not os.path.exists(dstdir):
+            os.makedirs(dstdir)
+        copy_file_path = os.path.join(dstdir, tmp_file_name)
+
+        self._download_file(download_type, file_url, copy_file_path)
+
+        return copy_file_path
+
+
+    def _download_file(self, download_type, file_url, copy_file_path):
```

Comment:
This and its subfunctions seem generally useful enough that they should live in DataFileUtils